File: chess/src/gameLogic/pieces/knight.js
import Piece from "./piece";
import Coord from "../helpers/coordinates";

export default class Knight extends Piece {
  constructor(...args) {
    super(args);
    this.type = "N";
  }

  copy() {
    return new Knight(...super.getInfos());
  }

  computeMoves(pieces, needToVerify) {
    this.moves.erase();
    const x = this.coord.x;
    const y = this.coord.y;

    const checkMove = (coord) => {
      if (coord.isValid()) {
        let otherPiece = pieces.findByCoord(coord);
        if (otherPiece) {
          if (otherPiece.team !== this.team) {
            this.moves.add(this, coord, "X", otherPiece);
          }
        } else this.moves.add(this, coord, "M", null);
      }
    };

    checkMove(new Coord(x - 1, y - 2));
    checkMove(new Coord(x - 2, y - 1));
    checkMove(new Coord(x + 1, y - 2));
    checkMove(new Coord(x + 2, y - 1));
    checkMove(new Coord(x - 1, y + 2));
    checkMove(new Coord(x - 2, y + 1));
    checkMove(new Coord(x + 1, y + 2));
    checkMove(new Coord(x + 2, y + 1));

    if (needToVerify) super.verifyMoves(pieces);
  }
}

File: engine/GameLogic/Pieces/Knight.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using ChessEngine.GameLogic.Utils;

namespace ChessEngine.GameLogic.Pieces
{
    public sealed class Knight : Piece
    {
        public Knight(bool team, Coord coord)
            : base(team, coord) { }

        public override void ComputeMoves(Dictionary<Coord, Piece> teamPieces, Dictionary<Coord, Piece> opponentPieces, bool needToVerify)
        {

            void Square(Coord coord)
            {
                if (coord.IsValid())
                {
                    if (opponentPieces.ContainsKey(coord))
                        Moves.Add(coord, new Move('X', coord));
                    else if (!teamPieces.ContainsKey(coord))
                        Moves.Add(coord, new Move('M', coord));
                }
            }

            Square(new Coord(Coord.x - 1, Coord.y - 2));
            Square(new Coord(Coord.x - 2, Coord.y - 1));
            Square(new Coord(Coord.x + 1, Coord.y - 2));
            Square(new Coord(Coord.x + 2, Coord.y - 1));
            Square(new Coord(Coord.x - 1, Coord.y + 2));
            Square(new Coord(Coord.x - 2, Coord.y + 1));
            Square(new Coord(Coord.x + 1, Coord.y + 2));
            Square(new Coord(Coord.x + 2, Coord.y + 1));

            if (needToVerify) base.VerifyMoves(teamPieces, opponentPieces);
        }
    }
}
